File: Binary Tree/二叉树的属性/112.路径总和.cpp
using namespace std;
/*
 * @lc app=leetcode.cn id=112 lang=cpp
 *
 * [112] 路径总和
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    bool getSum(TreeNode *root, int result)
    {
        bool left = false;
        bool right = false;

        // 叶子结点，同时和相等，满足条件
        if (root->val == result && root->left == nullptr && root->right == nullptr)
        {
            return true;
        }

        // 减去当前节点的值，剩下的和
        result = result - root->val;
        if (root->left != nullptr)
        {
            left = getSum(root->left, result);
        }
        if (root->right != nullptr)
        {
            right = getSum(root->right, result);
        }

        return left || right;
    }
    bool hasPathSum(TreeNode *root, int targetSum)
    {
        if (root == nullptr)
        {
            return false;
        }

        return getSum(root, targetSum);
    }
};
// @lc code=end

File: Binary Tree/二叉树的属性/112.路径总和.ts
/*
 * @lc app=leetcode.cn id=112 lang=typescript
 *
 * [112] 路径总和
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function hasPathSum(root: TreeNode | null, targetSum: number): boolean {
  const getSum = (sum: number, node: TreeNode | null) => {
    if (!node) {
      return -1
    }

    if (!node.left && !node.right) {
      if (node.val === sum) {
        return 0
      } else {
        return -1
      }
    }

    let result = sum - node.val
    let left = getSum(result, node.left)
    if (left === 0) {
      return 0
    }

    let right = getSum(result, node.right)
    if (right === 0) {
      return 0
    }

    return -1
  }

  return getSum(targetSum, root) === 0
};
// @lc code=end

